File: src/edu/vanderbilt/vandyvans/services/ReminderService.java
package edu.vanderbilt.vandyvans.services;

import java.lang.ref.WeakReference;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.util.Log;

import edu.vanderbilt.vandyvans.R;
import edu.vanderbilt.vandyvans.models.ArrivalTime;
import edu.vanderbilt.vandyvans.models.Stops;

import static edu.vanderbilt.vandyvans.services.Global.APP_LOG_ID;

/**
 * The Service that queries the Syncromatics server for arrival times.
 *
 * When a van arrives, publish the result and create a Notification.
 *
 * Created by athran on 4/6/14.
 */
public class ReminderService extends Service implements Handler.Callback {

    public static final int SUBSCRIBE_TO_STOP   = 42;
    public static final int UNSUBSCRIBE_TO_STOP = 43;
    public static final int VAN_IS_ARRIVEN      = 44;
    public static final int ADD_CLIENT          = 45;
    public static final int REMOVE_CLIENT       = 46;

    private static final int STOP_TRACKING = 47;
    private static final int INIT          = 48;

    private HandlerThread workerThread;
    private Messenger     clientHandler;
    private Handler       trackerHandler;
    private Handler       syncroHandler;

    private WeakReference<Messenger> mSingleClient = null;
    private Set<StopTracker>         mTrackers     = new HashSet<StopTracker>();
    private int                      mReminderId   = 1;

    @Override
    public void onCreate() {
        logMessage("Service is created");
        workerThread = new HandlerThread("reminderbackgroundthread");
        workerThread.start();

        syncroHandler  = new Handler(workerThread.getLooper(), new SyncromaticsClient());
        clientHandler  = new Messenger(new Handler(this));
        trackerHandler = new Handler(this);
    }

    @Override
    public IBinder onBind(Intent intent) {
        logMessage("Service is bound");
        return clientHandler.getBinder();
    }

    @Override
    public void onDestroy() {
        logMessage("Service is destroyed");
        workerThread.quit();
    }

    @Override
    public boolean handleMessage(Message message) {
        switch (message.what) {
            case SUBSCRIBE_TO_STOP:
                logMessage("Received subscription request: " +  message.arg1);
                StopTracker tracker = new StopTracker(message.arg1,
                                                      trackerHandler,
                                                      syncroHandler);
                Message.obtain(tracker, INIT)
                        .sendToTarget();
                mTrackers.add(tracker);
                break;

            case UNSUBSCRIBE_TO_STOP:
                logMessage("Received unsubscription request: " +  message.arg1);
                StopTracker toBeStopped = null;
                for (StopTracker trkr : mTrackers) {
                    if (trkr.stopId == message.arg1) {
                        toBeStopped = trkr;
                    }
                }

                if (toBeStopped != null) {
                    mTrackers.remove(toBeStopped);
                    Message.obtain(toBeStopped, STOP_TRACKING)
                            .sendToTarget();
                }
                break;

            case ADD_CLIENT:
                // keep just a single client for now
                mSingleClient = new WeakReference<Messenger>(message.replyTo);
                break;

            case REMOVE_CLIENT:
                mSingleClient = null;
                break;

            case VAN_IS_ARRIVEN:
                StopTracker reportingTracker = (StopTracker)message.obj;
                mTrackers.remove(reportingTracker);
                doBroadcastVanArrival(reportingTracker.latestArrivalTime);

                if (mTrackers.isEmpty()) { stopSelf(); }
                break;
        }
        return true;
    }

    private void doBroadcastVanArrival(ArrivalTime arrivalTime) {
        Messenger client = mSingleClient.get();
        if (client != null) {
            try {
                client.send(Message.obtain(null,
                                           VAN_IS_ARRIVEN,
                                           arrivalTime.stop.id,
                                           0));
                logMessage("notified client");
            } catch (RemoteException e) {
                logMessage("Failed to communicate with the client");
            }
        }

        NotificationManager notificationManager =
                (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        logMessage("About to broadcast for: " + arrivalTime.toString());
        mReminderId++;
        notificationManager.notify(
                mReminderId,
                new Notification.Builder(this)
                        .setContentTitle("Van Arriving")
                        .setContentText(new StringBuilder("The ")
                                                .append(arrivalTime.route.name)
                                                .append(" Route will be arriving at ")
                                                .append(arrivalTime.stop.name)
                                                .append(" in 5 minutes")
                                                .toString())
                        .setSmallIcon(R.drawable.van_icon)
                        .getNotification());
    }

    private static void logMessage(String msg) {
        Log.i(APP_LOG_ID, "ReminderService | " + msg);
    }

    private static final class StopTracker extends Handler {
        final int     stopId;
        final Handler syncroClient;
        final Handler parent;

        boolean     isTracking;
        ArrivalTime latestArrivalTime;

        private static final int NOTIFY_PARENT = 55;

        StopTracker(int     _id,
                    Handler _parent,
                    Handler _syncroClient) {
            super();
            stopId       = _id;
            parent       = _parent;
            syncroClient = _syncroClient;

            isTracking = true;
        }

        @Override public void handleMessage(Message msg) {
            if (msg.what == INIT) {
                logMessage("Initing for stopd: " + stopId);
                isTracking = true;
                Message.obtain(syncroClient, 0,
                               new Global.FetchArrivalTimes(this,
                                                            Stops.getForId(stopId)))
                        .sendToTarget();
            }
            if (msg.what == STOP_TRACKING) {
                isTracking = false;
            }
            if (msg.what == NOTIFY_PARENT) {
                if (isTracking) {
                    logMessage("Notifying parent");
                    Message.obtain(parent,
                                   VAN_IS_ARRIVEN,
                                   this).sendToTarget();
                }
            }
            if (msg.obj instanceof Global.ArrivalTimeResults) {
                handleArrivalTimes(((Global.ArrivalTimeResults) msg.obj).times);
            }
        }

        private void handleArrivalTimes(List<ArrivalTime> times) {
            logMessage("Received this many ArrivalTimes: " + times.size());
            for (ArrivalTime time : times) {
                logMessage(time.toString());
            }

            if (times.isEmpty()) { return; }

            latestArrivalTime = Collections.min(times, new Comparator<ArrivalTime>() {
                @Override
                public int compare(ArrivalTime lhs, ArrivalTime rhs) {
                    return lhs.minutes - rhs.minutes;
                }
            });

            if (latestArrivalTime.minutes > 5) {
                logMessage("Scheduling delayed notification");
                logMessage(latestArrivalTime.toString());
                sendMessageDelayed(Message.obtain(this, NOTIFY_PARENT),
                                   (latestArrivalTime.minutes - 5) * 60000);
            } else {
                logMessage("Scheduling immediate notification");
                logMessage(latestArrivalTime.toString());
                Message.obtain(this, NOTIFY_PARENT).sendToTarget();
            }
        }

    }

}

File: app/src/main/scala/edu/vanderbilt/vandyvans/services/ReminderService.scala
package edu.vanderbilt.vandyvans.services

import android.app.{Notification, NotificationManager, Service}
import android.content.{Context, Intent}
import android.os._
import android.util.Log
import com.marsupial.eventhub.ActorConversion
import edu.vanderbilt.vandyvans.models.{Stops, ArrivalTime}

import scala.ref.WeakReference

class ReminderServices extends Service with Handler.Callback {
  import ReminderServices._

  lazy val workerThread = new HandlerThread("reminderThread")
  lazy val syncroHandler = new Handler(workerThread.getLooper, new SyncromaticsClient)
  lazy val clientHandler = new Messenger(new Handler(this))
  lazy val trackerHandler = new Handler(this)

  var singleClient = Option.empty[WeakReference[Messenger]]
  var trackers = Set.empty[StopTracker]
  var reminderId = 1

  override def onCreate() {
    logMessage("Service is created")
    workerThread.start()
  }

  override def onBind(intent: Intent) = {
    logMessage("Service is bound")
    clientHandler.getBinder
  }

  override def onDestroy() {
    logMessage("Service is destroyed")
    workerThread.quit()
  }

  override def handleMessage(msg: Message) = {
    msg.what match {
      case SUBSCRIBE_TO_STOP =>
        logMessage(s"Received subscription request: ${message.arg1}")
        val tracker = new StopTracker(message.arg1, trackerHandler, syncroHandler)
        Message.obtain(tracker, INIT).sendToTarget()
        trackers = trackers + tracker

      case UNSUBSCRIBE_TO_STOP =>
        logMessage("Received unsubscription request: " + message.arg1)
        trackers.find(_.id == message.arg1).foreach { tracker =>
          Message.obtain(tracker, STOP_TRACKING).sendToTarget()
          trackers = trackers - tracker
        }

      case ADD_CLIENT =>
        singleClient = Some(WeakReference(msg.replyTo))

      case REMOVE_CLIENT =>
        singleClient = None

      case VAN_IS_ARRIVEN =>
        val reportingTracker = msg.obj.asInstanceOf[StopTracker]
        trackers = tracker - reportingTracker
        doBroadcastVanArrival(reportingTracker.latestArrivalTime)
        if (trackers.isEmpty) { stopSelf() }
    }
  }

  private def doBroadcastVanArrival(arrivalTime: ArrivalTime) {
    for (clientRef <- singleClient;
         client <- clientRef.get) {
      try {
        client.send(Message.obtain(null, VAN_IS_ARRIVEN, arrivalTime.stop.id, 0))
        logMessage("Notified client")
      } catch {
        case _: RemoteException => logMessage("Failed to communicate with the client")
      }
    }

    val notificationMan = getSystemService(Context.NOTIFICATION_SERVICE).asInstanceOf[NotificationManager]
    logMessage(s"About to broadcast for: ${arrivalTime.toString}")
    reminderId = reminderId + 1
    notificationMan.notify(
      reminderId,
      new Notification.Builder(this)
        .setContentTitle("Van Arriving")
        .setContentText(
          s"The ${arrivalTime.route.name} Route will be arriving at " +
          s"${arrivalTime.stop.name} in 5 minutes")
        .setSmallIcon(R.drawable.van_icon)
        .build())
  }

}

object ReminderServices {
  val SUBSCRIBE_TO_STOP: Int = 42
  val UNSUBSCRIBE_TO_STOP: Int = 43
  val VAN_IS_ARRIVEN: Int = 44
  val ADD_CLIENT: Int = 45
  val REMOVE_CLIENT: Int = 46
  private val STOP_TRACKING: Int = 47
  private val INIT: Int = 48
  val NOTIFY_PARENT = 55

  def logMessage(msg: String) {
    Log.i(Global.APP_LOG_ID, "ReminderService | " + msg)
  }

  class StopTracker(val id: Int,
                    val parent: Handler,
                    val syncro: Handler)
    extends Handler with ActorConversion
  {
    var isTracking = true
    var latestArrivalTime = Option.empty[ArrivalTime]

    override def handleMessage(msg: Message) {
      if (msg.what == INIT) {
        logMessage("Initing for stopId: " + stopId)
        isTracking = true
        syncro ! new SyncromaticsClient.FetchArrivalTimes(this, Stops.getForId(id))

      } else if (msg.what == NOTIFY_PARENT) {
        if (isTracking) {
          logMessage("Notifying parent")
          Message.obtain(parent, VAN_IS_ARRIVEN, this).sendToTarget()
        }

      } else if (msg.obj.isInstanceOf[SyncromaticsClient.ArrivalTimeResults]) {
        handleArrivalTimes(msg.obj.times)
      }
    }

    private def handleArrivalTimes(times: Iterable[ArrivalTime]) {
      logMessage("Received this many ArrivalTimes: " + times.size)
      times.foreach { t => logMessage(t.toString) }

      if (times.nonEmpty) {
        latestArrivalTime = Some(times.toSeq.sortBy(_.minutes).head)
        for (latest <- latestArrivalTime) {
          if (latest.minutes > 5) {
            logMessage("Scheduling delayed notification")
            logMessage(latestArrivalTime.toString)
            sendMessageDelayed(Message.obtain(this, NOTIFY_PARENT), (latestArrivalTime.minutes - 5) * 60000)
          } else {
            logMessage("Scheduling immediate notification")
            logMessage(latestArrivalTime.toString)
            Message.obtain(this, NOTIFY_PARENT).sendToTarget()
          }
        }
      }
    }
  }
}